src/CodeGenerator/AutoloadGenerator.php

```diff
@@ -7,87 +7,94 @@
-            . '}'.PHP_EOL
-            . 'return $this;';
+    private function buildFromTemplate(string $templateFile, string $outputFile, array $replacements) : void
+    {
+        $template = file_get_contents($templateFile);
+        $code = strtr($template, $replacements);
+        $outputFile = $this->outputDirectory . '/' . $outputFile;
 
-        $unregisterCode = 'if ($this->registered) {'.PHP_EOL
-            . '    spl_autoload_unregister($this);'.PHP_EOL
-            . '    $this->registered = false;'.PHP_EOL
-            . '}'.PHP_EOL
-            . 'return $this;';
+        $this->writeFile($outputFile, $code);
+    }
 
-        $loadCode = 'if (isset($this->classmap[$class])) {'.PHP_EOL
-            . '    include __DIR__ . \'/\' . $this->classmap[$class];'.PHP_EOL
-            . '}';
+    private function generateClassmapCode(array &$classmap) : string
+    {
+        $lines = array_map(
+            function (string $class, string $file): string {
+                return sprintf(
+                    '%s => %s,',
+                    var_export($class, true),
+                    var_export($file, true)
+                );
+            },
+            array_keys($classmap),
+            $classmap
+        );
 
-        $class
-            ->addProperty('registered', false, PropertyGenerator::FLAG_PRIVATE)
-            ->addProperty('classmap', $classmapValue, PropertyGenerator::FLAG_PRIVATE)
-            ->addMethod('register', [], MethodGenerator::FLAG_PUBLIC, $registerCode)
-            ->addMethod('unregister', [], MethodGenerator::FLAG_PUBLIC, $unregisterCode)
-            ->addMethod('load', ['class'], MethodGenerator::FLAG_PUBLIC, $loadCode)
-            ->addMethod('__invoke', ['class'], MethodGenerator::FLAG_PUBLIC, '$this->load($class);');
+        $indent = str_repeat(' ', 8);
+        return implode("\n$indent", $lines);
+    }
 
-        $file = new FileGenerator();
-        $file
-            ->setDocBlock(new DocBlockGenerator('Generated autoloader for Zend\Di'))
-            ->setNamespace($this->namespace)
-            ->setClass($class)
-            ->setFilename($this->outputDirectory . '/Autoloader.php');
+    private function generateAutoloaderClass(array &$classmap) : void
+    {
+        $this->buildFromTemplate(self::CLASS_TEMPLATE, 'Autoloader.php', [
+            '%namespace%' => $this->namespace ? sprintf("namespace %s;\n", $this->namespace) : '',
+            '%classmap%' => $this->generateClassmapCode($classmap),
+        ]);
+    }
 
-        $file->write();
+    private function generateAutoloadFile() : void
+    {
+        $this->buildFromTemplate(self::FILE_TEMPLATE, 'autoload.php', [
+            '%namespace%' => $this->namespace ? sprintf("namespace %s;\n", $this->namespace) : '',
+        ]);
     }
 
     /**
-     * @param array $classmap
+     * @param string[] $classmap
      */
-    public function generate(array &$classmap)
+    public function generate(array $classmap) : void
```

Comment:
The change from by-ref to by-val is a subtle BC break

Reply:
The implementation never changed the input and I'd consider extending this class is an edge case. In fact I'd like to make it final. But if it's a big deal this can be changed back to by ref. But I'd consider this confusing if the input is not modified.

Reply:
Still a BC break, even if the implementation didn't change. You can keep the by-ref, and we create a separate issue to remove it in the next major

Reply:
Sure